agent/stats/queue.go

```diff
@@ -186,6 +190,70 @@ func (queue *Queue) GetStorageStatsSet() (*ecstcs.StorageStatsSet, error) {
 	return storageStatsSet, errOut
 }
 
+// GetRestartStatsSet gets the stats set for container restarts
+func (queue *Queue) GetRestartStatsSet() (*ecstcs.RestartStatsSet, error) {
+	return queue.getRestartStatsSet(getRestartCount)
+}
+
+func (queue *Queue) getRestartStatsSet(getInt getIntPointerFunc) (*ecstcs.RestartStatsSet, error) {
+	queue.lock.Lock()
+	defer queue.lock.Unlock()
+
+	var firstStat, lastStat int64
+	firstStat = -1
+
+	queueLength := len(queue.buffer)
+	if queueLength < 2 {
+		// Need at least 2 data points to calculate this.
+		return nil, fmt.Errorf("need at least 2 data points in queue to calculate CW stats set")
+	}
+
+	for i, stat := range queue.buffer {
+		if stat.sent {
+			// don't send stats to TACS if already sent
+			continue
+		}
+		thisStat := getInt(&stat)
+		if thisStat == nil {
+			continue
+		}
+
+		if firstStat == -1 {
+			// firstStat is unset so set it here
+			if i > 0 {
+				// some stats in queue were sent already, so use the stat previous to the
+				// first to diff with the last stat.
+				thisStat = getInt(&queue.buffer[i-1])
+				if thisStat == nil {
+					continue
+				}
+			}
+			firstStat = *thisStat
+		}
+		if i == queueLength-1 {
+			// get final unsent stat in the queue
+			lastStat = *thisStat
+		}
+	}
+
+	if firstStat == -1 {
+		// no non-nil stats found, most likely this means that the container
+		// does not have a restart policy set.
+		return nil, fmt.Errorf("No non-nil restart count stats found, not sending RestartStatsSet." +
+			" Most likely the container does not have a restart policy configured")
+	}
+
+	// get the diff in restart count between the first stat and the last stat in the
+	// queue. Examples:
+	//    [ 0 1 3 3 4 5 ] = 5 - 0 = 5 restarts
+	//    [ 0(sent) 1(sent) 3(sent) 4(unsent) 5(unsent) 7(unsent) ] = 7 - 3 = 4 restarts
```

Comment:
can this be tested with a unit/integ test?

Reply:
yes this is tested in TestContainerStatsCollection_WithRestartPolicy, in that unit test the stats are verified, then the queue is reset, which sets the metrics to "sent" status, and then adds more metrics and again verifies the stats.